npep/npep-122.md

```diff
@@ -0,0 +1,182 @@
+and currently Tenancy is defined on the peers side. Tenancy by itself has the same subject and peer, 
+at least for the existing use cases, and having separate selectors for subject and peer allows for more
+configurations than needed.
+
+## API
+
+We suggest to define a separate API Object for tenancy, as in the following example:
+
+```
+kind: AdminNetworkPolicyTenancy
+apiVersion: policy.networking.k8s.io/v1alpha1
+metadata:
+  name: user-based
+spec:
+  namespaceSelector:
+    matchExpression:
+      - key: system-namespace
+        operator: DoesNotExist
+      - key: user
+        operator: Exists
+  tenancyLabels:
+    - user
+  priority: 99
+  ingress:
+    - action: Deny
+      from:
+        - NotSameTenant
+```
+
+that will define a tenant as follows:
+- first, select namespaces that will be split into tenants, and have the defined policy applied. `spec.namespaceSelector`
+is used for this purpose. This also solves the problem we had for empty or unset labels: if tenancy is defined based on label `user`,
+and namespaces without this label should not be affected, `key: user, operator: Exists` selector can be used.
+- second, selected namespaces are split into tenants based on label values, `spec.tenancyLabels` field is used for that purpose.
+Tenants with the same listed label values will form a tenant. Unset label and empty string label are also values, therefore
+in case namespaces with these label values are selected, they will be split into a separate tenant.
+
+    > Example: say we define a tenancy based on `label1` and `label2`, and the following namespaces are selected by `spec.namespaceSelector`
+    > ```
+    > ns1{label1=A, label2=B, label3=C}
+    > ns2{label1=A, label2=B, label4=D}
+    > ns3{label1=_, label2=B}
+    > ns4{label1=_, label2=""}
+    > ns5{label1=_, label2=_}
+    > ns6{label1=A, label2=B}
+    > ```
+    > where `_` means the label is not set, and `""` means empty string value.
+    >
+    > The following tenants will be created for `tenancyLabels: [label1, label2]`
+    > ```
+    > tenant{label1=A, label2=B} = [ns1, ns2, ns6]
+    > tenant{label1=_, label2=B} = [ns3]
+    > tenant{label1=_, label2=""} = [ns4]
+    > tenant{label1=_, label2=_} = [ns5]
+    > ```
+
+- third, define policy rules. This includes setting ports, action, and peer tenants. 
+For now, we only have `SameTenant` and `NotSameTenant` as a peer option.
+
+**Note** with the new tenancy definition there is no need to have `sameLabels` and `notSameLabels`, because
+these selectors will define the same tenants, but it is replaces with `SameTenant` and `NotSameTenant` for similar use cases.
+
+Existing API allows defining tenancy for BANP, we need to decide if the tenancy actually makes sense for that priority.
+BANP is overridable by the namespace owners, so the question is if there is a use case for a tenancy policy that can be
+overridden?
```

Comment:
There is. There are two kinds of "tenant" situation. One is the "Coke vs Pepsi" sort of thing where you want each tenant to feel like it has its own cluster, and be totally independent from the other tenants. In that case you want hard/non-overridable tenancy. But the other case is the "Engineering vs Marketing" or "OpenShift team vs RHEL team" sort of tenancy, where you want to keep people from _accidentally_ interfering with each other, but you want to allow them to communicate with each other if they want to. So for that you want isolated-by-default, but overridable.

(I think as multi-cluster becomes better supported, and things like HyperShift allow you to share resources between multiple clusters, there will be fewer people trying to do hard tenancy within a single cluster.)

Reply:
I find these use cases very easy to understand, so I added user stories clarifications section